x/dsmr/partition.go

```diff
@@ -0,0 +1,195 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package dsmr
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils"
+
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/internal/cache"
+)
+
+const partitionCacheSize = 100
+
+type weightedValidator struct {
+	weight uint64
+	nodeID ids.NodeID
+}
+
+type PrecalculatedPartition[T Tx] struct {
+	validators  []*weightedValidator
+	totalWeight uint64
+}
+
+func (w *weightedValidator) Compare(o *weightedValidator) int {
+	return bytes.Compare(w.nodeID[:], o.nodeID[:])
+}
+
+func CalculatePartition[T Tx](ctx context.Context, state validators.State, pChainHeight uint64, subnetID ids.ID) (*PrecalculatedPartition[T], error) {
+	vdrs, err := state.GetValidatorSet(ctx, pChainHeight, subnetID)
+	if err != nil {
+		return nil, err
+	}
+	weightedValidators := make([]*weightedValidator, 0, len(vdrs))
+	totalWeight := uint64(0)
+	for _, vdr := range vdrs {
+		weightedValidators = append(weightedValidators, &weightedValidator{
+			weight: vdr.Weight,
+			nodeID: vdr.NodeID,
+		})
+		totalWeight += vdr.Weight
+	}
+	utils.Sort(weightedValidators)
+
+	return &PrecalculatedPartition[T]{
+		validators:  weightedValidators,
+		totalWeight: totalWeight,
+	}, nil
+}
+
+func calculateSponsorWeightIndex(sponsor codec.Address, totalWeight uint64) uint64 {
+	return binary.BigEndian.Uint64(sponsor[len(sponsor)-consts.Uint64Len:]) % totalWeight
+}
+
+func (pp *PrecalculatedPartition[T]) AssignTx(tx T) (ids.NodeID, bool) {
```

Comment:
the implementation of this function is not very efficient.
instead, in `PrecalculatedPartition` we want to have an array called
``accumulatedTotalWeights` where each elements would contain the total accumulated weight until that validator.
Then, in AssignTx, we want to perform a binary search ( O(logn) ) instead of array scan ( O(n) ).

Reply:
Great idea